File: packages/rank/ios/Sources/RankPlugin/RankImpl.swift
import Foundation
import StoreKit

/**
 * Native iOS implementation for the Rank plugin.
 *
 * This class contains pure platform logic and is isolated from the Capacitor bridge.
 * Architectural constraints:
 * - MUST NOT access CAPPluginCall.
 * - MUST NOT depend on Capacitor bridge APIs directly.
 * - MUST perform UI operations on the Main Thread.
 */
@objc public final class RankImpl: NSObject {

    // MARK: - Properties

    /// Cached plugin configuration containing logging and behavioral flags.
    private var config: RankConfig?

    // MARK: - Initialization

    /**
     * Initializes the implementation instance.
     */
    override init() {
        super.init()
    }

    // MARK: - Configuration

    /**
     * Applies static plugin configuration.
     *
     * This method MUST be called exactly once from the Plugin bridge layer during `load()`.
     * It synchronizes the native logger state with the provided configuration.
     *
     * - Parameter config: The immutable configuration container.
     */
    func applyConfig(_ config: RankConfig) {
        precondition(
            self.config == nil,
            "RankImpl.applyConfig(_:) must be called exactly once"
        )
        self.config = config
        RankLogger.verbose = config.verboseLogging

        RankLogger.debug(
            "Configuration applied. Verbose logging:",
            config.verboseLogging
        )
    }

    // MARK: - Availability

    /**
     * Checks if the current iOS version supports the SKStoreReviewController API.
     * Since the plugin target is iOS 15+, this generally returns true.
     */
    @objc public func isAvailable() -> Bool {
        // In-App Review is supported on iOS 15+ (minimum deployment target).
        return true
    }

    // MARK: - Product Page

    /**
     * Presents the App Store product page within the application.
     * * - Parameter appId: The numeric Apple App ID.
     * - Parameter completion: Callback to signal when the page has been loaded or failed.
     */
    @objc public func presentProductPage(appId: String, completion: @escaping (Bool, Error?) -> Void) {
        let storeViewController = SKStoreProductViewController()
        let parameters = [SKStoreProductParameterITunesItemIdentifier: appId]

        // UI operations must be on the main thread
        DispatchQueue.main.async {
            storeViewController.loadProduct(withParameters: parameters) { success, error in
                if success {
                    // Locate the active key window in a multi-scene environment.
                    let rootViewController = UIApplication.shared.connectedScenes
                        .compactMap { $0 as? UIWindowScene }
                        .flatMap { $0.windows }
                        .first { $0.isKeyWindow }?
                        .rootViewController

                    guard let rootVC = rootViewController else {
                        // No active root view controller available for presentation.
                        completion(false, nil)
                        return
                    }

                    rootVC.present(storeViewController, animated: true) {
                        completion(true, nil)
                    }
                } else {
                    completion(false, error)
                }
            }
        }
    }

    // MARK: - Review Methods

    /**
     * Triggers the native iOS In-App Review prompt.
     *
     * Uses the modern `requestReview(in:)` API for iOS 14+ by identifying the active foreground scene.
     * Fallback logic is provided for older iOS versions (though the plugin target is iOS 15+).
     *
     * NOTE: The OS manages the frequency of this prompt; calling this does not guarantee a UI will appear.
     */
    @objc public func requestReview() {
        // All UI-related operations must be executed on the main thread.
        DispatchQueue.main.async {
            // In-App Review is available on iOS 15+ (minimum deployment target).
            let activeScene = UIApplication.shared.connectedScenes
                .compactMap { $0 as? UIWindowScene }
                .first { $0.activationState == .foregroundActive }

            guard let windowScene = activeScene else {
                // Review prompt cannot be requested without an active foreground scene.
                RankLogger.error("Cannot request review: no active UIWindowScene found")
                return
            }

            RankLogger.debug("Requesting in-app review for active UIWindowScene")
            SKStoreReviewController.requestReview(in: windowScene)
        }
    }

    // MARK: - Store Navigation

    /**
     * Opens the App Store page for the specified application ID.
     *
     * This method constructs a deep link using the `itms-apps` scheme to direct the user
     * to the review section of the application's store page.
     *
     * - Parameter appId: The numeric Apple App ID (e.g., "123456789").
     */
    @objc public func openStore(appId: String) {
        guard let url = RankUtils.appStoreURL(appId: appId) else {
            RankLogger.error("Invalid App Store URL for ID: \(appId)")
            return
        }

        // RULE: All UI-related operations MUST be dispatched to the main thread.
        DispatchQueue.main.async {
            UIApplication.shared.open(url, options: [:]) { success in
                if success {
                    RankLogger.debug("App Store opened successfully.")
                } else {
                    RankLogger.error("Failed to open App Store URL.")
                }
            }
        }
    }

    /**
     * Opens the App Store product page for the specified application ID.
     *
     * This method constructs a deep link using the `itms-apps` scheme to direct the user
     * to the product page of the application in the App Store.
     *
     * - Parameter appId: The numeric Apple App ID (e.g., "123456789").
     */
    @objc public func openStoreListing(appId: String) {
        let urlString = "itms-apps://itunes.apple.com/app/id\(appId)"
        if let url = URL(string: urlString) {
            DispatchQueue.main.async {
                UIApplication.shared.open(url)
            }
        }
    }

    /**
     * Opens the App Store search page with the specified search terms.
     *
     * This method constructs a deep link using the `itms-apps` scheme to direct the user
     * to the search results page in the App Store for the given terms.
     *
     * - Parameter terms: The search query string to find relevant apps in the App Store.
     */
    @objc public func search(terms: String) {
        guard let url = RankUtils.searchURL(terms: terms) else { return }
        DispatchQueue.main.async {
            UIApplication.shared.open(url)
        }
    }
}

File: packages/rank/android/src/main/java/io/capkit/rank/RankImpl.kt
package io.capkit.rank

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import com.google.android.play.core.review.ReviewManagerFactory
import io.capkit.rank.error.RankErrorMessages
import io.capkit.rank.utils.RankLogger
import io.capkit.rank.utils.RankUtils
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Platform-specific native implementation for the Rank plugin.
 *
 * This class contains pure Android logic and MUST NOT depend directly on
 * Capacitor bridge APIs or PluginCall objects.
 *
 * Responsibilities:
 * - Orchestrating the Google Play Review SDK.
 * - Managing Intent-based Store navigation.
 * - Translating configuration into native behavior.
 */
class RankImpl(
  private val context: Context,
) {
  data class ReviewEnvironmentDiagnostic(
    val canRequestReview: Boolean,
    val reason: String? = null,
    val error: RankError? = null,
  )

  // ---------------------------------------------------------------------------
  // Properties
  // ---------------------------------------------------------------------------

  /**
   * Cached plugin configuration container.
   * Provided once during initialization via [updateConfig].
   */
  private lateinit var config: RankConfig

  /**
   * Prevents concurrent review flow executions.
   */
  private val isReviewInProgress = AtomicBoolean(false)

  // ---------------------------------------------------------------------------
  // Configuration
  // ---------------------------------------------------------------------------

  /**
   * Applies the plugin configuration to the implementation layer.
   *
   * This method MUST be called exactly once during the plugin [RankPlugin.load]
   * phase. It initializes internal state and configures logging verbosity.
   *
   * @param newConfig The immutable configuration instance.
   */
  fun updateConfig(newConfig: RankConfig) {
    this.config = newConfig
    RankLogger.verbose = newConfig.verboseLogging
    RankLogger.debug(
      "Configuration applied. Verbose logging:",
      newConfig.verboseLogging.toString(),
    )
  }

  // ---------------------------------------------------------------------------
  // Availability check
  // ---------------------------------------------------------------------------

  /**
   * Checks whether the In-App Review feature is available on this device.
   *
   * This is a semantic availability check intended for UI decisions.
   * Internally, it delegates to the diagnostic environment check.
   */
  fun isAvailable(onResult: (Boolean) -> Unit) {
    checkReviewEnvironment { diagnostic ->
      onResult(diagnostic.canRequestReview)
    }
  }

  /**
   * Performs a diagnostic check for the Google Play Review environment.
   *
   * This verifies whether Google Play Services can provide
   * a ReviewInfo instance for the current application and device.
   *
   * No UI is triggered by this operation.
   */
  fun checkReviewEnvironment(onResult: (ReviewEnvironmentDiagnostic) -> Unit) {
    val playStorePackage = "com.android.vending"
    val playStoreIntent =
      Intent(Intent.ACTION_VIEW, RankUtils.marketDetailsUri(context.packageName)).apply {
        setPackage(playStorePackage)
      }

    if (playStoreIntent.resolveActivity(context.packageManager) == null) {
      onResult(
        ReviewEnvironmentDiagnostic(
          canRequestReview = false,
          reason = "PLAY_STORE_NOT_AVAILABLE",
        ),
      )
      return
    }

    val installerPackage =
      try {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.R) {
          context.packageManager.getInstallSourceInfo(context.packageName).installingPackageName
        } else {
          @Suppress("DEPRECATION")
          context.packageManager.getInstallerPackageName(context.packageName)
        }
      } catch (_: PackageManager.NameNotFoundException) {
        null
      } catch (_: Exception) {
        null
      }

    if (installerPackage != playStorePackage) {
      onResult(
        ReviewEnvironmentDiagnostic(
          canRequestReview = false,
          reason = "NOT_INSTALLED_FROM_PLAY_STORE",
        ),
      )
      return
    }

    val manager = ReviewManagerFactory.create(context)
    val request = manager.requestReviewFlow()

    request.addOnCompleteListener { task ->
      if (task.isSuccessful) {
        onResult(ReviewEnvironmentDiagnostic(canRequestReview = true))
      } else {
        onResult(
          ReviewEnvironmentDiagnostic(
            canRequestReview = false,
            error =
              RankError.Unavailable(
                task.exception?.message ?: RankErrorMessages.PLAY_CORE_REVIEW_API_UNAVAILABLE,
              ),
          ),
        )
      }
    }
  }

  // ---------------------------------------------------------------------------
  // Pre-warm Logic
  // ---------------------------------------------------------------------------

  /**
   * Cached ReviewInfo object to speed up the review flow display.
   */
  private var cachedReviewInfo: com.google.android.play.core.review.ReviewInfo? = null

  /**
   * Pre-fetches the ReviewInfo from Google Play Services.
   * This should be called early (e.g., during plugin load).
   */
  fun preloadReviewInfo() {
    val manager = ReviewManagerFactory.create(context)
    val request = manager.requestReviewFlow()
    request.addOnCompleteListener { task ->
      if (task.isSuccessful) {
        cachedReviewInfo = task.result
        RankLogger.debug("ReviewInfo pre-loaded successfully.")
      } else {
        RankLogger.debug("ReviewInfo pre-load failed.")
      }
    }
  }

  // ---------------------------------------------------------------------------
  // In-App Review Logic
  // ---------------------------------------------------------------------------

  /**
   * Triggers the Google Play In-App Review flow.
   * Uses the cached ReviewInfo if available to ensure an immediate prompt.
   *
   * @param activity The current Android Activity required to display the UI.
   * @param onComplete Callback invoked when the flow completes or fails.
   */
  fun requestReview(
    activity: Activity,
    onComplete: (Exception?) -> Unit,
  ) {
    if (!isReviewInProgress.compareAndSet(false, true)) {
      onComplete(IllegalStateException("Review flow already in progress"))
      return
    }

    val manager = ReviewManagerFactory.create(context)

    // Use cache if available, otherwise fetch new info on the fly
    if (cachedReviewInfo != null) {
      val flow = manager.launchReviewFlow(activity, cachedReviewInfo!!)
      flow.addOnCompleteListener { _ ->
        cachedReviewInfo = null // Clear cache after use to prevent reuse of expired info
        isReviewInProgress.set(false)
        onComplete(null)
      }
    } else {
      val request = manager.requestReviewFlow()
      request.addOnCompleteListener { task ->
        if (task.isSuccessful) {
          val reviewInfo = task.result
          val flow = manager.launchReviewFlow(activity, reviewInfo)
          flow.addOnCompleteListener { _ ->
            isReviewInProgress.set(false)
            onComplete(null)
          }
        } else {
          isReviewInProgress.set(false)
          onComplete(task.exception)
        }
      }
    }
  }

  // ---------------------------------------------------------------------------
  // Store Navigation Logic
  // ---------------------------------------------------------------------------

  /**
   * Opens the Google Play Store page for the specified package.
   *
   * It attempts to use the "market://" URI scheme first to open the native
   * Play Store app. If the app is missing, it falls back to a browser-based
   * "https://" URL.
   *
   * @param packageName The application package name. Defaults to the host app if null.
   */
  fun openStore(packageName: String?) {
    val targetPackage = packageName ?: context.packageName

    // Construct the native market Intent
    val intent =
      Intent(Intent.ACTION_VIEW, RankUtils.marketDetailsUri(targetPackage)).apply {
        addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
      }

    try {
      context.startActivity(intent)
    } catch (e: Exception) {
      // Fallback to web browser if Play Store application is unavailable
      val webIntent =
        Intent(
          Intent.ACTION_VIEW,
          RankUtils.playStoreDetailsHttpsUri(targetPackage),
        ).apply {
          addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        }
      context.startActivity(webIntent)
    }
  }

  /**
   * Opens the Google Play Store listing for a specific application ID.
   *
   * @param appId The Android package name or application ID to display.
   */
  fun openStoreListing(appId: String) {
    val intent =
      Intent(Intent.ACTION_VIEW, RankUtils.marketDetailsUri(appId)).apply {
        addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
      }

    try {
      context.startActivity(intent)
    } catch (e: Exception) {
      val webIntent =
        Intent(
          Intent.ACTION_VIEW,
          RankUtils.playStoreDetailsHttpsUri(appId),
        ).apply {
          addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        }
      context.startActivity(webIntent)
    }
  }

  /**
   * Launches the Google Play Store search results for the provided terms.
   *
   * @param terms The search query string.
   */
  fun search(terms: String) {
    val intent =
      Intent(Intent.ACTION_VIEW, RankUtils.marketSearchUri(terms)).apply {
        addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
      }
    context.startActivity(intent)
  }

  /**
   * Navigates to a specific developer's page on the Google Play Store.
   *
   * @param devId The unique developer identifier (numeric or string ID).
   */
  fun openDevPage(devId: String) {
    val intent =
      Intent(Intent.ACTION_VIEW, RankUtils.playStoreDeveloperUri(devId)).apply {
        addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
      }
    context.startActivity(intent)
  }

  /**
   * Opens a curated collection or category on the Google Play Store.
   *
   * @param name The collection identifier (e.g., "featured", "editors_choice").
   */
  fun openCollection(name: String) {
    val intent =
      Intent(Intent.ACTION_VIEW, RankUtils.playStoreCollectionUri(name)).apply {
        addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
      }
    context.startActivity(intent)
  }
}
